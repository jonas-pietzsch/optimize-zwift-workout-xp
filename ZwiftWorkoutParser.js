import { XMLParser, XMLBuilder } from "fast-xml-parser";
import { ZwiftWorkout } from "./ZwiftWorkout.js";

const options = {
  ignoreAttributes: false,
  attributeNamePrefix: "attribute",
  preserveOrder: true,
};

export class ZwiftWorkoutParser {
  static parseZwiftWorkoutFile(fileBuffer) {
    const parser = new XMLParser(options);
    const parsedWorkoutFile = parser.parse(fileBuffer.toString());
    return new ZwiftWorkout(parsedWorkoutFile);
  }

  static assembleZwiftWorkoutFile(zwiftWorkout) {
    const builder = new XMLBuilder({
      ...options,
      format: true,
    });
    return builder.build(zwiftWorkout.contents);
  }
}
