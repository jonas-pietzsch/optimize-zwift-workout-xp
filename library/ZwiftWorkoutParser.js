import { XMLParser, XMLBuilder } from "fast-xml-parser";
import { ZwiftWorkout } from "./ZwiftWorkout.js";

const options = {
  ignoreAttributes: false,
  attributeNamePrefix: "attribute",
  preserveOrder: true,
};

export class ZwiftWorkoutParser {
  static xmlParser = new XMLParser(options);
  static xmlBuilder = new XMLBuilder({ ...options, format: true });

  static parseZwoFile(fileBuffer) {
    const parsedWorkoutFile = ZwiftWorkoutParser.xmlParser.parse(
      fileBuffer.toString()
    );
    return new ZwiftWorkout(parsedWorkoutFile);
  }

  static assembleZwoFile(zwiftWorkout) {
    return ZwiftWorkoutParser.xmlBuilder.build(zwiftWorkout.contents);
  }
}
