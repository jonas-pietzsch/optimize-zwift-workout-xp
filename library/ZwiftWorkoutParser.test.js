import { ZwiftWorkoutParser } from "./ZwiftWorkoutParser.js";
import * as fs from "fs";

describe("ZwiftWorkoutParser", () => {
  it("should parse zwo file contents to an object structure that preserves sorting, type and attribute of workout segments", () => {
    const workout = ZwiftWorkoutParser.parseZwiftWorkoutFile(
      fs.readFileSync("./sample.zwo")
    );
    expect(workout.contents).toMatchSnapshot();
  });

  it("should re-create the same zwo file contents when parsing it and generating it again without changes", () => {
    const originalFileBuffer = fs.readFileSync("./sample.zwo");
    const originalFileContentsWithoutWhitespaces = originalFileBuffer
      .toString()
      .replace(/\s/g, "");
    const workout =
      ZwiftWorkoutParser.parseZwiftWorkoutFile(originalFileBuffer);

    const targetFileContents =
      ZwiftWorkoutParser.assembleZwiftWorkoutFile(workout);
    const targetFileContentsWithoutWhitespaces = targetFileContents.replace(
      /\s/g,
      ""
    );

    expect(originalFileContentsWithoutWhitespaces).toEqual(
      targetFileContentsWithoutWhitespaces.replace("<tags></tags>", "<tags/>")
    );

    const workoutFromReGeneratedZwoFile =
      ZwiftWorkoutParser.parseZwiftWorkoutFile(Buffer.from(targetFileContents));

    expect(workout.contents).toEqual(workoutFromReGeneratedZwoFile.contents);
  });
});
