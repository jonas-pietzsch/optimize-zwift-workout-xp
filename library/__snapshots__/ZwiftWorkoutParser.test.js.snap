// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ZwiftWorkoutParser should parse zwo file contents to an object structure that preserves sorting, type and attribute of workout segments 1`] = `
[
  {
    "workout_file": [
      {
        "author": [
          {
            "#text": "Somebody",
          },
        ],
      },
      {
        "name": [
          {
            "#text": "Something",
          },
        ],
      },
      {
        "description": [
          {
            "#text": "Some description",
          },
        ],
      },
      {
        "sportType": [
          {
            "#text": "bike",
          },
        ],
      },
      {
        "tags": [],
      },
      {
        "workout": [
          {
            ":@": {
              "attributeDuration": "920",
              "attributePowerHigh": "0.65",
              "attributePowerLow": "0.45",
            },
            "Warmup": [],
          },
          {
            ":@": {
              "attributeDuration": "20",
              "attributePowerHigh": "1",
              "attributePowerLow": "0.7",
            },
            "Warmup": [],
          },
          {
            ":@": {
              "attributeCadence": "70",
              "attributeDuration": "180",
              "attributePower": "0.6",
            },
            "SteadyState": [],
          },
          {
            ":@": {
              "attributeCadence": "80",
              "attributeDuration": "180",
              "attributePower": "0.7",
            },
            "SteadyState": [],
          },
          {
            ":@": {
              "attributeCadence": "90",
              "attributeDuration": "180",
              "attributePower": "0.8",
            },
            "SteadyState": [],
          },
          {
            ":@": {
              "attributeCadence": "100",
              "attributeDuration": "180",
              "attributePower": "0.9",
            },
            "SteadyState": [],
          },
          {
            ":@": {
              "attributeCadence": "100",
              "attributeDuration": "20",
              "attributePower": "1.2",
            },
            "SteadyState": [],
          },
          {
            ":@": {
              "attributeCadence": "85",
              "attributeDuration": "300",
              "attributePower": "0.55",
            },
            "SteadyState": [],
          },
          {
            ":@": {
              "attributeOffDuration": "30",
              "attributeOffPower": "0.53",
              "attributeOnDuration": "30",
              "attributeOnPower": "1.33",
              "attributeRepeat": "10",
            },
            "IntervalsT": [],
          },
          {
            ":@": {
              "attributeDuration": "300",
              "attributePower": "0.55",
            },
            "SteadyState": [],
          },
          {
            ":@": {
              "attributeOffDuration": "30",
              "attributeOffPower": "0.53",
              "attributeOnDuration": "30",
              "attributeOnPower": "1.33",
              "attributeRepeat": "10",
            },
            "IntervalsT": [],
          },
          {
            ":@": {
              "attributeDuration": "300",
              "attributePower": "0.55",
            },
            "SteadyState": [],
          },
          {
            ":@": {
              "attributeOffDuration": "30",
              "attributeOffPower": "0.53",
              "attributeOnDuration": "30",
              "attributeOnPower": "1.33",
              "attributeRepeat": "10",
            },
            "IntervalsT": [],
          },
          {
            ":@": {
              "attributeDuration": "1080",
              "attributePowerHigh": "0.43",
              "attributePowerLow": "0.63",
            },
            "Cooldown": [],
          },
        ],
      },
    ],
  },
]
`;
